File: LeetCode/DP/UniqueBinarySearchTreesII.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {number} n
 * @return {TreeNode[]}
 */
var generateTrees = function(n) {
    
  if (n < 1) return [];
  const mapping = [...Array(n + 1)].map(ele => Array(n + 1));

  const buildTree = (low, high) => {
      const res = [];
      if (low > high) {
          res.push(null);
          return res;
      }

      if (mapping[low][high]) return mapping[low][high]

      for (let cnt = low; cnt <= high; cnt ++) {
          const leftSubTree = buildTree(low, cnt - 1);
          const rightSubTree = buildTree(cnt + 1, high);
          
          if (rightSubTree.length === 2) {
              console.log(rightSubTree);
              console.log("test")
              console.log(leftSubTree)
          }

          for (const left of leftSubTree) {
              for (const right of rightSubTree) {
                  const root = new TreeNode(cnt);
                  root.left = left;
                  root.right = right;
                  res.push(root);
              }
          }
      }

      mapping[low][high] = res;

      return res;
  }

  return buildTree(1, n);
};

File: LeetCode/DP/UniqueBinarySearchTreesII.java
import java.util.LinkedList;
import java.util.List;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

/**
 * @jingjiejiang Oct 24, 2019
 */
class Solution {
    public class TreeNode {
      int val;
      TreeNode left;
      TreeNode right;
      TreeNode(int x) { val = x; }
    }

    public List<TreeNode> generateTrees(int n) {
        List<TreeNode> list = new LinkedList<>();
        if (n == 0) return list;

        // use a hashmap : "low#high" -- List<TreeNode> for record previous status

        return genTreesHelper(1, n);
    }

    private List<TreeNode> genTreesHelper(int low, int high) {
        List<TreeNode> res = new LinkedList<>();  
        if (low > high) {
            res.add(null);
            return res;
        }

        for (int cnt = low; cnt <= high; cnt ++) {
            List<TreeNode> leftNodes = genTreesHelper(low, cnt - 1);
            List<TreeNode> rightNodes = genTreesHelper(cnt + 1, high);

            for (TreeNode leftNode : leftNodes) {
              for (TreeNode rightNode : rightNodes) {
                TreeNode root = new TreeNode(cnt);
                root.left = leftNode;
                root.right = rightNode;
                res.add(root);
              }
            }
        }

        return res;
    }
}